Implement CSS font-size-adjust property

vitorroriz opened the pull request:
#### 6fed42070a60c9c4eb396c7a552df09ea6878241
<pre>
Implement CSS font-size-adjust property
<a href="https://bugs.webkit.org/show_bug.cgi?id=15257">https://bugs.webkit.org/show_bug.cgi?id=15257</a>
rdar://23286433

Reviewed by Myles C. Maxfield.

Implementing font-size-adjust CSS Fonts Module level 4 [1].
[1]: <a href="https://www.w3.org/TR/css-fonts-4/#font-size-adjust-prop">https://www.w3.org/TR/css-fonts-4/#font-size-adjust-prop</a>

* LayoutTests/TestExpectations:
* LayoutTests/fast/css/getComputedStyle/computed-style-font-family-expected.txt:
* LayoutTests/imported/w3c/resources/resource-files.json:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/animations/font-size-adjust-interpolation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-005-expected.xht: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-012-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-012-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-012.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/inheritance-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-size-adjust-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-size-adjust-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/resources/ahem-ex-250.otf: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/resources/ahem-ex-500.otf: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-pseudo/first-letter-allowed-properties-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-pseudo/first-line-allowed-properties-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-pseudo/parsing/marker-supported-properties-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-pseudo/parsing/marker-supported-properties-in-animation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-variables/variable-presentation-attribute-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-detached-subtree-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-pseudo-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/styling/presentation-attributes-irrelevant-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/styling/presentation-attributes-relevant-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/styling/presentation-attributes-unknown-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/styling/required-properties-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/responsive/font-size-adjust-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/import/text-fonts-204-t-manual-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-003-expected.txt: Added.
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-detached-subtree-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-pseudo-expected.txt:
* LayoutTests/platform/gtk/tables/mozilla/bugs/bug2886-2-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-pseudo-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-003-expected.txt: Added.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-detached-subtree-expected.txt:
* LayoutTests/platform/ios/tables/mozilla/bugs/bug2886-2-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-detached-subtree-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-pseudo-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-003-expected.txt: Added.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-005-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug2886-2-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-detached-subtree-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-pseudo-expected.txt:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::CSSPrimitiveValue::CSSPrimitiveValue):
(WebCore::CSSPrimitiveValue::operator std::optional&lt;float&gt; const):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::fontSizeAdjustFromStyle):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::consumeFontSizeAdjust):
(WebCore::CSSPropertyParser::parseSingleValue):
* Source/WebCore/platform/graphics/FontCascadeCache.h:
(WebCore::FontDescriptionKeyRareData::create):
(WebCore::FontDescriptionKeyRareData::fontSizeAdjust const):
(WebCore::FontDescriptionKeyRareData::operator== const):
(WebCore::FontDescriptionKeyRareData::FontDescriptionKeyRareData):
(WebCore::add):
(WebCore::FontDescriptionKey::FontDescriptionKey):
* Source/WebCore/platform/graphics/FontCascadeDescription.cpp:
* Source/WebCore/platform/graphics/FontCascadeDescription.h:
(WebCore::FontCascadeDescription::initialFontSizeAdjust):
* Source/WebCore/platform/graphics/FontDescription.h:
(WebCore::FontDescription::fontSizeAdjust const):
(WebCore::FontDescription::setFontSizeAdjust):
(WebCore::FontDescription::operator== const):
(WebCore::FontDescription::encode const):
(WebCore::FontDescription::decode):
* Source/WebCore/platform/graphics/FontPlatformData.cpp:
(WebCore::FontPlatformData::cloneWithSize):
(WebCore::FontPlatformData::updateSize):
(WebCore::FontPlatformData::updateSizeWithFontSizeAdjust):
* Source/WebCore/platform/graphics/FontPlatformData.h:
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::FontCache::createFontPlatformData):
* Source/WebCore/platform/graphics/cocoa/FontFamilySpecificationCoreText.cpp:
(WebCore::FontFamilySpecificationCoreText::fontRanges const):
* Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
(WebCore::FontPlatformData::cloneWithSize):
(WebCore::FontPlatformData::updateSize):
* Source/WebCore/platform/graphics/freetype/FontCacheFreeType.cpp:
(WebCore::FontCache::createFontPlatformData):
* Source/WebCore/platform/graphics/freetype/FontCustomPlatformDataFreeType.cpp:
(WebCore::FontCustomPlatformData::fontPlatformData):
* Source/WebCore/platform/graphics/freetype/FontPlatformDataFreeType.cpp:
(WebCore::FontPlatformData::cloneWithSize):
(WebCore::FontPlatformData::updateSize):
* Source/WebCore/platform/graphics/freetype/SimpleFontDataFreeType.cpp:
(WebCore::Font::platformCreateScaledFont const):
* Source/WebCore/platform/graphics/mac/FontCustomPlatformData.cpp:
(WebCore::FontCustomPlatformData::fontPlatformData):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::setLetterSpacing):
(WebCore::RenderStyle::setFontSize):
(WebCore::RenderStyle::setFontSizeAdjust):
(WebCore::RenderStyle::setFontVariationSettings):
(WebCore::RenderStyle::setFontWeight):
(WebCore::RenderStyle::setFontStretch):
(WebCore::RenderStyle::setFontItalic):
(WebCore::RenderStyle::setFontPalette):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::fontSizeAdjust const):
* Source/WebCore/style/PropertyAllowlist.cpp:
(WebCore::Style::isValidMarkerStyleProperty):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueFontSizeAdjust):
* Source/WebCore/style/StyleFontSizeFunctions.cpp:
(WebCore::Style::adjustedFontSize):
* Source/WebCore/style/StyleFontSizeFunctions.h:
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::createAttributeNameToCSSPropertyIDMap):

Canonical link: <a href="https://commits.webkit.org/255927@main">https://commits.webkit.org/255927@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/592c1df837645c8fa4a3ea1219e091b1bef70ca9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94044 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3236 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/24593 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103669 "Updated wpe dependencies (failure)") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/164020 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3249 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31465 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86358 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/99720 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99704 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/2332 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80466 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/29353 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84258 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/83941 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/72306 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/37853 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/17773 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35726 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/19037 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4093 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39601 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/41615 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41540 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/38269 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -2325,6 +2325,38 @@ class PropertyWrapperFontStyle final : public PropertyWrapper<std::optional<Font
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    PropertyWrapperFontSizeAdjust()
+        : PropertyWrapperGetter<std::optional<float>>(CSSPropertyFontSizeAdjust, &RenderStyle::fontSizeAdjust) 
```
I believe for base classes that are template classes you can leave out the arguments in a constructor like this, so this can and should be:
```
    : PropertyWrapperGetter(CSSPropertyFontSizeAdjust, &RenderStyle::fontSizeAdjust)
```

darinadler commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -2325,6 +2325,38 @@ class PropertyWrapperFontStyle final : public PropertyWrapper<std::optional<Font
+                return (!context.progress ? from : to).fontSizeAdjust();
+            ASSERT(from.fontSizeAdjust().has_value() && to.fontSizeAdjust().has_value());
+            auto blendedAdjust = blendFunc(from.fontSizeAdjust().value(), to.fontSizeAdjust().value(), context);
+            return std::make_optional<float>(std::max(blendedAdjust, 0.0f));
```
Should not need to specify `<float>` here so this should be:
```
            return std::make_optional(std::max(blendedAdjust, 0.0f));
```
The other option is to use `-> std::optional<float>` when defining the lambda above, and then we could omit `make_optional` here.

litherum replied:
Huh. That's neat.

darinadler commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -2325,6 +2325,38 @@ class PropertyWrapperFontStyle final : public PropertyWrapper<std::optional<Font
+            return std::make_optional<float>(std::max(blendedAdjust, 0.0f));
+        };
+
+        auto* currentFontSelector = destination.fontCascade().fontSelector();
```
Not sure we need the word "current" in this local name. I would just name this `selector` or `fontSelector`.

But more importantly, this code is just a copy of the code in the `RenderStyle::setFontSizeAdjust` function so we should write: `destination.setFontSizeAdjust(blendedFontSizeAdjust())`.

darinadler commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -3199,6 +3231,7 @@ CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap()
 #if ENABLE(VARIATION_FONTS)
         new PropertyWrapperFontVariationSettings(CSSPropertyFontVariationSettings, &RenderStyle::fontVariationSettings, &RenderStyle::setFontVariationSettings),
 #endif
+        new PropertyWrapperFontSizeAdjust(),
```
No need for `()` here.

darinadler commented on Source/WebCore/css/CSSPrimitiveValueMappings.h:
```diff
@@ -5518,6 +5518,28 @@ template<> inline CSSPrimitiveValue::operator FontSynthesisLonghandValue() const
+    if (!isNumber())
+        return std::nullopt;
+
+    return std::make_optional<float>(floatValue());
```
This should just be:
```
    return floatValue();
```

darinadler commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -2200,6 +2200,12 @@ static Ref<CSSPrimitiveValue> fontSizeFromStyle(const RenderStyle& style)
+static Ref<CSSPrimitiveValue> fontSizeAdjustFromStyle(const RenderStyle& style)
+{
+    auto fontSizeAdjust = style.fontSizeAdjust();
+    return CSSValuePool::singleton().createValue(fontSizeAdjust);
```
No need for a local variable here, just one line:
```
    return CSSValuePool::singleton().createValue(style.fontSizeAdjust());
```

darinadler commented on Source/WebCore/platform/graphics/FontCascadeCache.h:
```diff
@@ -178,6 +180,7 @@ struct FontDescriptionKey {
     std::array<unsigned, 2> m_flags { { 0, 0 } };
     AtomString m_locale;
     RefPtr<FontDescriptionKeyRareData> m_rareData;
+    std::optional<float> m_fontSizeAdjust;
```
Not sure we need the word "font" here since the context is a structure named `FontDescriptionKey`. Maybe just `m_sizeAdjust`?

vitorroriz replied:
I wanted to avoid confusion with the size-adjust property, that we don't implement yet.

darinadler commented on Source/WebCore/platform/graphics/FontCascadeDescription.h:
```diff
@@ -136,6 +136,7 @@ class FontCascadeDescription : public FontDescription {
     static FontOpticalSizing initialOpticalSizing() { return FontOpticalSizing::Enabled; }
     static const AtomString& initialSpecifiedLocale() { return nullAtom(); }
     static FontPalette initialFontPalette() { return { FontPalette::Type::Normal, nullAtom() }; }
+    static std::optional<float> initialFontSizeAdjust() { return std::nullopt; }
```
I’d be tempted to use `constexpr` here. Does no harm, but maybe does no good either?

darinadler commented on Source/WebCore/platform/graphics/FontDescription.h:
```diff
@@ -46,7 +46,9 @@ class FontDescription {
 
     float computedSize() const { return m_computedSize; }
     unsigned computedPixelSize() const { return unsigned(m_computedSize + 0.5f); }
+    bool hasFontSizeAdjust() const { return m_sizeAdjust.has_value(); }
```
Same thought here. Do we need "font" in all the function names?

litherum replied:
Why wouldn't callers just call fontSizeAdjust() and convert to a bool context at that point?

darinadler commented on Source/WebCore/platform/graphics/FontPlatformData.cpp:
```diff
@@ -59,7 +63,10 @@ FontPlatformData FontPlatformData::cloneWithSyntheticOblique(const FontPlatformD
+#endif
 
+#if !USE(FREETYPE) && !USE(CORE_TEXT) 
+// fixme: don't other platforms also need to reinstantiate its copy.m_font for scaled size?
```
We use `FIXME`, not `fixme`. Capitalization mistake here: "Don't" should be capitalized. Grammar mistake here: for the subject and verb to agree it should be platforms/their or platform/its.

litherum replied:
we say `FIXME`

Better to include a link to a bug

darinadler commented on Source/WebCore/platform/graphics/FontPlatformData.h:
```diff
@@ -126,7 +126,11 @@ class FontPlatformData {
+
+    // cloneWitSize to be implemented by each platform since it needs to re-instantiate the platform font object.
     static FontPlatformData cloneWithSize(const FontPlatformData&, float);
+    static FontPlatformData cloneWithAdjustedSize(const FontPlatformData&, float);
```
Unfortunate that this has to copy the source when the size is already correct. Consider renaming and changing the argument to `FontPlatformData&&`? I understand this fits in with the existing clone functions, but most of those seem to copy unconditionally, so this is different.

One way to do this is to change this into a member function that modifies a `FontPlatformData` in place.

litherum replied:
I don't think this function needs to be public.

vitorroriz replied:
Good point. However, I've just realized `FontPlatformData` doesn't have a move constructor, which also makes the move in `fontPlatformDataWithAdjustedSize` irrelevant (also since it is returning I'll remove explicitly move there).

 Should we have a move ctor for it or there's a reason for not having one? @litherum

vitorroriz replied:
I've implemented the in-place version and modified cloneWithSize functions to use it.

darinadler commented on Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
```diff
@@ -778,6 +779,7 @@ RetainPtr<CTFontRef> preparePlatformFont(CTFontRef originalFont, const FontDescr
 
     addAttributesForInstalledFonts(attributes.get(), fontDescription.shouldAllowUserInstalledFonts());
 
+    // todo: font-size-adjust: I can maybe add adjusted size here instead
```
This project uses `FIXME`, not `todo`. Also, it would be good to have an easier to understand comment here. Like "can maybe add" is pretty ambiguous. Could you write down why you think it might be needed to work correctly?

vitorroriz replied:
tldr: Removing it.

It's a note to myself I forgot to remove. I was wondering if we could adjust the font-size at that point together with other alterations. This is unfortunately not possible because we need Font metrics, thus we would need a FontPlatformData object at this point.

darinadler commented on Source/WebCore/platform/graphics/cocoa/FontFamilySpecificationCoreText.cpp:
```diff
@@ -46,8 +47,8 @@ FontFamilySpecificationCoreText::~FontFamilySpecificationCoreText() = default;
-    auto& fontPlatformData = FontFamilySpecificationCoreTextCache::forCurrentThread().ensure(FontFamilySpecificationKey(m_fontDescriptor.get(), fontDescription), [&] () {
-        auto size = fontDescription.computedSize();
+    auto size = fontDescription.computedSize();
+    FontPlatformData platformData = FontFamilySpecificationCoreTextCache::forCurrentThread().ensure(FontFamilySpecificationKey(m_fontDescriptor.get(), fontDescription), [&] () {
```
Fix the formatting here to `[&]()` without the space, please.

litherum replied:
(FYI https://webkit.org/code-style-guidelines/#spacing-lambda-paren)

litherum replied:
Previously it says `auto& fontPlatformData` and now it says `FontPlatformData platformData`

1. I think it's better to use `auto` because the name tells you the type
2. Making an unconditional copy of the FontPlatformData is unfortunate. It's sizable.

darinadler commented on Source/WebCore/platform/graphics/cocoa/FontFamilySpecificationCoreText.cpp:
```diff
@@ -57,7 +58,8 @@ FontRanges FontFamilySpecificationCoreText::fontRanges(const FontDescription& fo
     });
-    return FontRanges(FontCache::forCurrentThread().fontForPlatformData(fontPlatformData));
+
+    return FontRanges(FontCache::forCurrentThread().fontForPlatformData(FontPlatformData::fontPlatformDataWithAdjustedSize(WTFMove(platformData), fontDescription)));
```
The explicit conversion to `FontRanges` was not needed before and is still not needed after.

litherum replied:
Same as above, I think we shouldn't make 2 FontPlatformDatas here.

vitorroriz replied:
FontRanges has a ctor that takes RefPtr<Font> (instead of Ref<Font>) but it is explicit.

darinadler commented on Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
```diff
@@ -394,6 +394,8 @@ bool Font::supportsAllPetiteCaps() const
 #endif
 
+
+// todo: should we adjust font for font-size-adjust in derivative font case?
```
We use FIXME, not todo, in this project, and we capitalize the sentence, "FIXME: Should [...]".

litherum replied:
Usually these kind of things are accompanied by a bug link.

darinadler commented on Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
```diff
@@ -394,6 +394,8 @@ bool Font::supportsAllPetiteCaps() const
 }
 #endif
 
+
```
Please don’t add an extra blank line here.

darinadler commented on Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
```diff
@@ -183,4 +184,16 @@ String FontPlatformData::familyName() const
+    FontPlatformData copy(source);
+    copy.m_size = size;
+    // We need to reinitialize the instance, because the difference in size
+    // necessitates a new scaled font instance.
```
Comment should be on one line. I don’t understand why we need this comment. Clearly we are allocating a new `CTFont` because one of the attributes is different, and clearly that is the size. If you think it’s not obvious, then I suppose we could write a comment: "Need a new CTFont with the new size". I would not use the phrase "reinitialize the instance", or the word "necessitates".

darinadler commented on Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
```diff
@@ -183,4 +184,16 @@ String FontPlatformData::familyName() const
+    // We need to reinitialize the instance, because the difference in size
+    // necessitates a new scaled font instance.
+    ASSERT(copy.m_font.get());
+    copy.m_font = adoptCF(CTFontCreateCopyWithAttributes(copy.m_font.get(), copy.m_size, nullptr, CTFontCopyFontDescriptor(copy.m_font.get())));
```
This leaks the result of `CTFontCopyFontDescriptor`. We have to use `adoptCF` to prevent that. But also pretty sure we don’t need to pass the font descriptor from the font itself with all the attributes the font already has. The descriptor is only needed for additional attributes that we want to be different from the ones already in the font. So we should just be passing `nullptr`.

We also have to set `copy.m_ctFont` to nullptr.

litherum replied:
Agreed, I think we should try to implement the `font-size-adjust` property without cloning any fonts, and without creating any additional font objects.

litherum commented on LayoutTests/platform/mac/tables/mozilla/bugs/bug2886-2-expected.txt:
```diff
@@ -3,105 +3,105 @@ layer at (0,0) size 800x600
 layer at (0,0) size 800x270
   RenderBlock {HTML} at (0,0) size 800x270
     RenderBody {BODY} at (88,25) size 648x220 [bgcolor=#FFFFF5]
-      RenderTable {TABLE} at (0,0) size 364x83
```
Why are there changes here? Does this test use the new property?

vitorroriz replied:
We've talked about this one offline. But just for registering it, it is styled by https://github.com/WebKit/WebKit/blob/main/LayoutTests/tables/mozilla/bugs/bug2886-2.css which uses font-size-adjust.

darinadler commented on Source/WebCore/platform/graphics/freetype/SimpleFontDataFreeType.cpp:
```diff
@@ -192,7 +192,7 @@ RefPtr<Font> Font::platformCreateScaledFont(const FontDescription& fontDescripti
     return Font::create(FontPlatformData(cairo_scaled_font_get_font_face(m_platformData.scaledFont()),
         m_platformData.fcPattern(),
-        scaleFactor * fontDescription.computedSize(),
+        (scaleFactor * m_platformData.size()),
```
Please don’t add parentheses here.

darinadler commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -2077,6 +2078,16 @@ void RenderStyle::setFontSize(float size)
 
+void RenderStyle::setFontSizeAdjust(std::optional<float> sizeAdjust)
+{
+    FontSelector* currentFontSelector = fontCascade().fontSelector();
```
`auto* selector` would be better

litherum replied:
Bonus points if you fix up the sibling functions above/below to do the same.

darinadler commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -1937,6 +1937,7 @@ const FontCascadeDescription& RenderStyle::fontDescription() const
     return m_inheritedData->fontCascade.fontDescription();
 }
 
+
```
Please don’t add this blank line.

darinadler commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -1910,6 +1911,17 @@ inline void BuilderCustom::applyValueFontSize(BuilderState& builderState, CSSVal
+    auto& primitiveValue = downcast<CSSPrimitiveValue>(value);
+
+    if (primitiveValue.isNumber() && !primitiveValue.isNegative().value_or(true)) {
+        fontDescription.setFontSizeAdjust(std::make_optional(primitiveValue.floatValue()));
```
The call to `std::make_optional` should not be needed here.

darinadler commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -1910,6 +1911,17 @@ inline void BuilderCustom::applyValueFontSize(BuilderState& builderState, CSSVal
+    auto fontDescription = builderState.fontDescription();
+    auto& primitiveValue = downcast<CSSPrimitiveValue>(value);
+
+    if (primitiveValue.isNumber() && !primitiveValue.isNegative().value_or(true)) {
```
Is a negative number the only kind of invalid number? I’m sort of surprised. I don’t see `isNegative` checks in other style builder functions. Is this sort of unique in that respect?

vitorroriz replied:
It can be non-negative or none for level 4. Maybe others are trusting that the parser will make the validation before, which we can also trust. But anyway there is a mistake in this code, we are not applying the value if it is a CSSValueNone. I'll fix it.

darinadler commented on Source/WebCore/style/StyleFontSizeFunctions.cpp:
```diff
@@ -183,5 +184,19 @@ int legacyFontSizeForPixelSize(int pixelFontSize, bool shouldUseFixedDefaultSize
 
+float adjustedFontSize(float size, float sizeAdjust, float xHeight)
+{
+    float aspectValue = xHeight / size;
```
What guarantees `size` is not zero?

What guarantees `xHeight` is not zero?

litherum replied:
`size` can definitely be 0, this is common in practice

darinadler commented on Source/WebCore/style/StyleFontSizeFunctions.h:
```diff
@@ -39,6 +40,8 @@ enum class MinimumFontSizeRule : uint8_t { None, Absolute, AbsoluteAndRelative }
 float computedFontSizeFromSpecifiedSize(float specifiedSize, bool isAbsoluteSize, float zoomFactor, MinimumFontSizeRule, const Settings::Values&);
 float computedFontSizeFromSpecifiedSize(float specifiedSize, bool isAbsoluteSize, bool useSVGZoomRules, const RenderStyle*, const Document&);
 float computedFontSizeFromSpecifiedSizeForSVGInlineText(float specifiedSize, bool isAbsoluteSize, float zoomFactor, const Document&);
+float adjustedFontSize(float size, float sizeAdjust, float xHeight);
```
I suggest making this overload of `adjustedFontSize` be private and not declared in the header.

darinadler reviewed (changes requested):
Looking good, some improvements I have suggested and some small mistakes that must be fixed.

litherum commented on Source/WebCore/css/CSSPrimitiveValueMappings.h:
```diff
@@ -5518,6 +5518,28 @@ template<> inline CSSPrimitiveValue::operator FontSynthesisLonghandValue() const
+    : CSSValue(PrimitiveClass)
+{
+    if (!sizeAdjust.has_value()) {
+        setPrimitiveUnitType(CSSUnitType::CSS_VALUE_ID);
+        m_value.valueID = CSSValueNone;
```
There's gotta be a way to do this more elegantly.

darinadler replied:
Hate to break it to you: there is not! But we could make one.

litherum commented on Source/WebCore/css/CSSPrimitiveValueMappings.h:
```diff
@@ -5518,6 +5518,28 @@ template<> inline CSSPrimitiveValue::operator FontSynthesisLonghandValue() const
+        return;
+    }
+
+    setPrimitiveUnitType(CSSUnitType::CSS_NUMBER);
+    m_value.num = static_cast<double>(sizeAdjust.value());
```
Ditto.

darinadler replied:
Same.

litherum commented on Source/WebCore/platform/graphics/FontCascadeCache.h:
```diff
@@ -126,7 +127,8 @@ struct FontDescriptionKey {
             && m_locale == other.m_locale
-            && arePointingToEqualData(m_rareData, other.m_rareData);
+            && arePointingToEqualData(m_rareData, other.m_rareData)
+            && m_fontSizeAdjust == other.m_fontSizeAdjust;
```
Why is font-size-adjust not considered rare?

litherum commented on Source/WebCore/platform/graphics/FontCascadeDescription.cpp:
```diff
@@ -43,13 +43,14 @@ struct SameSizeAsFontCascadeDescription {
     unsigned bitfields3 : 10;
 };
-
+template<int s> struct Wow;
```
???

vitorroriz replied:
That's actually funny. It was for investigating a bit alignment issue. Forgot to remove everything, sorry.

litherum commented on Source/WebCore/platform/graphics/FontPlatformData.h:
```diff
@@ -126,7 +126,11 @@ class FontPlatformData {
     static FontPlatformData cloneWithOrientation(const FontPlatformData&, FontOrientation);
     static FontPlatformData cloneWithSyntheticOblique(const FontPlatformData&, bool);
+
+    // cloneWitSize to be implemented by each platform since it needs to re-instantiate the platform font object.
```
`Wit` typo

litherum commented on Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
```diff
@@ -1342,7 +1343,9 @@ std::unique_ptr<FontPlatformData> FontCache::createFontPlatformData(const FontDe
-    return makeUnique<FontPlatformData>(font.get(), size, syntheticBold, syntheticOblique, fontDescription.orientation(), fontDescription.widthVariant(), fontDescription.textRenderingMode());
+    FontPlatformData platformData(font.get(), size, syntheticBold, syntheticOblique, fontDescription.orientation(), fontDescription.widthVariant(), fontDescription.textRenderingMode());
+
+    return makeUnique<FontPlatformData>(FontPlatformData::fontPlatformDataWithAdjustedSize(WTFMove(platformData), fontDescription));
```
I think this is the wrong way to do it. Let's not make a whole FontPlatformData, just to throw it away and make another slightly different one. Let's calculate what the difference is, and then just pass that difference into the first constructor. (The difference is just some simple math on floats.)

litherum replied:
We discussed offline. The approach I suggested doesn't directly work, because we can't get the x-height without a CTFont. Even if we refactored things to do it this way, the logic to populate the x-height in FontMetrics is non-trivial. And even if we compensated for that, the `override` descriptors in `@font-face` would require more custom stuff on top, which we'd have to remember to do.... The approach taken in this patch is probably better.

litherum commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -368,9 +368,11 @@ class RenderStyle {
     WEBCORE_EXPORT const FontCascade& fontCascade() const;
     WEBCORE_EXPORT const FontMetrics& metricsOfPrimaryFont() const;
     WEBCORE_EXPORT const FontCascadeDescription& fontDescription() const;
+    float fontSize() const;
```
I think we shouldn't add this. I think it's a feature that the client has to pick whether they want the specified or computed one. The relationship between them is somewhat complicated due to iOS text autosizing.

litherum commented on Source/WebCore/style/StyleFontSizeFunctions.cpp:
```diff
@@ -183,5 +184,19 @@ int legacyFontSizeForPixelSize(int pixelFontSize, bool shouldUseFixedDefaultSize
+float adjustedFontSize(float size, float sizeAdjust, float xHeight)
+{
+    float aspectValue = xHeight / size;
+    return size * ( sizeAdjust / aspectValue );
```
let's remove the spaces just inside the parentheses.

litherum commented on Source/WebCore/style/StyleFontSizeFunctions.cpp:
```diff
@@ -183,5 +184,19 @@ int legacyFontSizeForPixelSize(int pixelFontSize, bool shouldUseFixedDefaultSize
+float adjustedFontSize(float size, float sizeAdjust, const FontMetrics& metrics)
+{
+    if (!metrics.hasXHeight())
+        return size;
```
Under which situations would the metrics not have an x height?

instead of saying the x height may not exist, can we just guarantee at setting-time that if it wouldn't exist, we synthesize a value (maybe 60% of font size)?

litherum commented on Source/WebCore/style/StyleFontSizeFunctions.cpp:
```diff
@@ -183,5 +184,19 @@ int legacyFontSizeForPixelSize(int pixelFontSize, bool shouldUseFixedDefaultSize
+}
+
 }
 }
```
I know you didn't write this but this might be a good place to add closing-brace comments.

E.g.
```
} // namespace Foo
```

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-012-expected.txt:
```diff
@@ -0,0 +1,48 @@
+layer at (0,0) size 800x600
```
Tests that have a rendertree dump as reference can't be exported to WPT.

You need another HTML file as reference, or the test has to be using the JS testharness.

vitorroriz commented on LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-012-ref.html:
```diff
@@ -0,0 +1,68 @@
+<html>
+    <meta charset="utf-8">
+    <head>
+        <link rel="match" href="font-size-adjust-012-ref.html">
```
ops, forgot to remove the link to ref from the ref itself. Will remove.

litherum commented on Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
```diff
@@ -183,4 +184,25 @@ String FontPlatformData::familyName() const
+    // copy.m_size = size;
+    // ASSERT(copy.m_font.get());
+    // copy.m_font = adoptCF(CTFontCreateCopyWithAttributes(copy.m_font.get(), copy.m_size, nullptr, nullptr));
+    // copy.m_ctFont = nullptr;
```
We don't generally commit commented-out code.

litherum commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -2043,11 +2043,11 @@ void RenderStyle::setWordSpacing(Length&& value)
 void RenderStyle::setLetterSpacing(float letterSpacing)
 {
-    FontSelector* currentFontSelector = fontCascade().fontSelector();
+    FontSelector* selector = fontCascade().fontSelector();
```
might as well use `auto`

litherum commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -2068,63 +2068,73 @@ void RenderStyle::setFontSize(float size)
         size = std::min(maximumAllowedFontSize, size);
 
-    FontSelector* currentFontSelector = fontCascade().fontSelector();
+    FontSelector* selector = fontCascade().fontSelector();
```
ditto

litherum commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -2068,63 +2068,73 @@ void RenderStyle::setFontSize(float size)
 
     setFontDescription(WTFMove(description));
-    fontCascade().update(currentFontSelector);
+    fontCascade().update(selector);
```
This is fairly substantial. Small drive-by's like this are usually OK, but this much is a bit distracting.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-012.html:
```diff
@@ -0,0 +1,73 @@
+<!DOCTYPE html>
+<html>
```
Can you match this file, the -ref.html, the -expected.html to the ones from your WPT PR?

vitorroriz replied:
ops, I did but forgot the -expected. Thank you! Fixed.

vitorroriz committed:
Implement CSS font-size-adjust property
https://bugs.webkit.org/show_bug.cgi?id=15257
rdar://23286433

Reviewed by Myles C. Maxfield.

Implementing font-size-adjust CSS Fonts Module level 4 [1].
[1]: https://www.w3.org/TR/css-fonts-4/#font-size-adjust-prop

* LayoutTests/TestExpectations:
* LayoutTests/fast/css/getComputedStyle/computed-style-font-family-expected.txt:
* LayoutTests/imported/w3c/resources/resource-files.json:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/animations/font-size-adjust-interpolation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-005-expected.xht: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-012-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-012-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-012.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/inheritance-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-size-adjust-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-size-adjust-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/resources/ahem-ex-250.otf: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/resources/ahem-ex-500.otf: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-pseudo/first-letter-allowed-properties-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-pseudo/first-line-allowed-properties-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-pseudo/parsing/marker-supported-properties-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-pseudo/parsing/marker-supported-properties-in-animation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-variables/variable-presentation-attribute-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-detached-subtree-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-pseudo-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/styling/presentation-attributes-irrelevant-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/styling/presentation-attributes-relevant-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/styling/presentation-attributes-unknown-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/styling/required-properties-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/responsive/font-size-adjust-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/import/text-fonts-204-t-manual-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-003-expected.txt: Added.
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-detached-subtree-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-pseudo-expected.txt:
* LayoutTests/platform/gtk/tables/mozilla/bugs/bug2886-2-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-pseudo-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-003-expected.txt: Added.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-detached-subtree-expected.txt:
* LayoutTests/platform/ios/tables/mozilla/bugs/bug2886-2-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-detached-subtree-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-pseudo-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-003-expected.txt: Added.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-005-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug2886-2-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-detached-subtree-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-pseudo-expected.txt:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::CSSPrimitiveValue::CSSPrimitiveValue):
(WebCore::CSSPrimitiveValue::operator std::optional<float> const):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::fontSizeAdjustFromStyle):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::consumeFontSizeAdjust):
(WebCore::CSSPropertyParser::parseSingleValue):
* Source/WebCore/platform/graphics/FontCascadeCache.h:
(WebCore::FontDescriptionKeyRareData::create):
(WebCore::FontDescriptionKeyRareData::fontSizeAdjust const):
(WebCore::FontDescriptionKeyRareData::operator== const):
(WebCore::FontDescriptionKeyRareData::FontDescriptionKeyRareData):
(WebCore::add):
(WebCore::FontDescriptionKey::FontDescriptionKey):
* Source/WebCore/platform/graphics/FontCascadeDescription.cpp:
* Source/WebCore/platform/graphics/FontCascadeDescription.h:
(WebCore::FontCascadeDescription::initialFontSizeAdjust):
* Source/WebCore/platform/graphics/FontDescription.h:
(WebCore::FontDescription::fontSizeAdjust const):
(WebCore::FontDescription::setFontSizeAdjust):
(WebCore::FontDescription::operator== const):
(WebCore::FontDescription::encode const):
(WebCore::FontDescription::decode):
* Source/WebCore/platform/graphics/FontPlatformData.cpp:
(WebCore::FontPlatformData::cloneWithSize):
(WebCore::FontPlatformData::updateSize):
(WebCore::FontPlatformData::updateSizeWithFontSizeAdjust):
* Source/WebCore/platform/graphics/FontPlatformData.h:
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::FontCache::createFontPlatformData):
* Source/WebCore/platform/graphics/cocoa/FontFamilySpecificationCoreText.cpp:
(WebCore::FontFamilySpecificationCoreText::fontRanges const):
* Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
(WebCore::FontPlatformData::cloneWithSize):
(WebCore::FontPlatformData::updateSize):
* Source/WebCore/platform/graphics/freetype/FontCacheFreeType.cpp:
(WebCore::FontCache::createFontPlatformData):
* Source/WebCore/platform/graphics/freetype/FontCustomPlatformDataFreeType.cpp:
(WebCore::FontCustomPlatformData::fontPlatformData):
* Source/WebCore/platform/graphics/freetype/FontPlatformDataFreeType.cpp:
(WebCore::FontPlatformData::cloneWithSize):
(WebCore::FontPlatformData::updateSize):
* Source/WebCore/platform/graphics/freetype/SimpleFontDataFreeType.cpp:
(WebCore::Font::platformCreateScaledFont const):
* Source/WebCore/platform/graphics/mac/FontCustomPlatformData.cpp:
(WebCore::FontCustomPlatformData::fontPlatformData):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::setLetterSpacing):
(WebCore::RenderStyle::setFontSize):
(WebCore::RenderStyle::setFontSizeAdjust):
(WebCore::RenderStyle::setFontVariationSettings):
(WebCore::RenderStyle::setFontWeight):
(WebCore::RenderStyle::setFontStretch):
(WebCore::RenderStyle::setFontItalic):
(WebCore::RenderStyle::setFontPalette):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::fontSizeAdjust const):
* Source/WebCore/style/PropertyAllowlist.cpp:
(WebCore::Style::isValidMarkerStyleProperty):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueFontSizeAdjust):
* Source/WebCore/style/StyleFontSizeFunctions.cpp:
(WebCore::Style::adjustedFontSize):
* Source/WebCore/style/StyleFontSizeFunctions.h:
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::createAttributeNameToCSSPropertyIDMap):

Canonical link: https://commits.webkit.org/255927@main
